pkg/ast/ast.go

```diff
@@ -202,6 +202,32 @@ func (d *Document) DeleteRootNode(node Node) {
 	}
 }
 
+// removeNodeAtIndexIgnoringOrder efficiently removes a node from a slice at a given index without preserving order.
+func removeNodeAtIndexIgnoringOrder(nodes []Node, indexToRemove int) []Node {
+	lastIndex := len(nodes) - 1
+	nodes[indexToRemove] = nodes[lastIndex]
+	return nodes[:lastIndex]
+}
+
+// DeleteRootNodesInSingleLoop loops over the document RootNodes exactly once. The RootNodes are deleted while looping,
+// so they are iterated over backwards; consequently, the ascending nodesToRemove of a single NodeKind are also iterated
+// over backwards to match the order they would appear among the RootNodes.
+func (d *Document) DeleteRootNodesInSingleLoop(rootNodesToRemove []Node) {
```

Comment:
The function name contains an implementation detail. Better to stick with `DeleteRootNotes`. 
I would also suggest adding some function-specific unit tests, because removing things can always lead to some bugs when not done right

Reply:
There is already a method called DeleteRootNodes with the same signature. :(